infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStorageProvisioner.java

```diff
@@ -0,0 +1,370 @@
+ * https://github.com/eclipse/che/issues/15384) This environment will allow backup on workspace stop
+ * event and restore on restart created earlier. <br>
+ * Will apply only in case workspace has attributes: asyncPersist: true - persistVolumes:
+ * false.</br> In case workspace has attributes: asyncPersist: true - persistVolumes: true will
+ * throw exception.</br> Feature enabled only for 'common' PVC strategy, in other cases will throw
+ * exception.</br> During provision will be created: - storage Pod - service for rsync connection
+ * via SSH - configmap, with public part of SSH key - PVC for storing backups;
+ */
+public class AsyncStorageProvisioner {
+
+  private static final int SERVICE_PORT = 2222;
+  private static final String AUTHORIZED_KEYS = "authorized_keys";
+  static final String ASYNC_STORAGE = "async-storage";
+  static final String ASYNC_STORAGE_CONFIG = "async-storage-config";
+  private static final String ASYNC_STORAGE_DATA_PATH = "/" + ASYNC_STORAGE;
+  private static final String SSH_KEY_PATH = "/.ssh/" + AUTHORIZED_KEYS;
+  static final String SSH_KEY_NAME = "rsync-via-ssh";
+  private static final String CONFIG_MAP_VOLUME_NAME = "async-storage-configvolume";
+  private static final String STORAGE_VOLUME = "async-storage-data";
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncStorageProvisioner.class);
+
+  private final String pvcQuantity;
+  private final String storageImage;
+  private final String accessMode;
+  private final String strategy;
+  private final String configuredPVCName;
+  private final String pvcStorageClassName;
+  private final SshManager sshManager;
+  private final OpenShiftClientFactory clientFactory;
+
+  @Inject
+  public AsyncStorageProvisioner(
+      @Named("che.infra.kubernetes.pvc.quantity") String pvcQuantity,
+      @Named("che.infra.kubernetes.async.storage.image") String image,
+      @Named("che.infra.kubernetes.pvc.access_mode") String accessMode,
+      @Named("che.infra.kubernetes.pvc.strategy") String strategy,
+      @Named("che.infra.kubernetes.pvc.name") String configuredPVCName,
+      @Named("che.infra.kubernetes.pvc.storage_class_name") String pvcStorageClassName,
+      SshManager sshManager,
+      OpenShiftClientFactory openShiftClientFactory) {
+    this.pvcQuantity = pvcQuantity;
+    this.storageImage = image;
+    this.accessMode = accessMode;
+    this.strategy = strategy;
+    this.configuredPVCName = configuredPVCName;
+    this.pvcStorageClassName = pvcStorageClassName;
+    this.sshManager = sshManager;
+    this.clientFactory = openShiftClientFactory;
+  }
+
+  public void provision(OpenShiftEnvironment osEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    if (!"true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))) {
+      return;
+    }
+
+    if ("true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))
+        && !"common".equals(strategy)) {
```

Comment:
CommonPVCStrategy.COMMON_STRATEGY

Reply:
Shouldn't this rather be a validation check during workspace creation?